src/Oscoin/Storage/BlockTree.hs

```diff
@@ -0,0 +1,74 @@
+-- | An opaque 'BlockTree', composed by its public and private API. Its main
+-- responsibilities are:
+--
+-- 1. Storing main blocks.
+-- 2. Storing orphan blocks.
+-- 3. Performing chain selection.
+-- 4. Perform full-chain validation.
+
+module Oscoin.Storage.BlockTree
+    ( BlockTree
+    , BlockTreeReader
+    , BlockTreeWriter
+
+    , hoistBlockTreeReader
+    , hoistBlockTreeWriter
```

Comment:
I have backported this hoist for simplicity but I hope we can avoid having this one in the first place, as ideally the less degree of freedom we expose when it comes to the writer, the better. We'll see how tests adjustment plays out.